PX4/Firmware
# towards systemwide replay

The following proposes a way to replay any part of the system, based on uORB topics. Replaying consists of 3 components:
- a replay module
- uORB publisher rules
- time handling

## replay module
This is a separate module that loads an ULog file, applies the parameters from it (and optionally applies user-supplied parameters from another file) and then publishes all topics it finds in 'real-time'. It also adds a constant offset to the timestamp of each message to match the current system time.
The log file name is supplied via an environment variable.

## uORB publisher rules
With a set of rules loaded from a file, the user can specify which part of the system should be replayed. In particular, the rules are fully integrated into orb and specify which module is allowed to publish which topics. This process is completely transparent for each publisher, so a publisher doesn't know (and doesn't need to know) that his messages are simply ignored. Here's an example:
```
restrict_topics: sensor_combined, vehicle_gps_position, vehicle_land_detected
module: replay
ignore_others:true
```
It means the rules should be applied for the given topics and the `replay` module. Only the `replay` module is allowed to publish these topics. The `ignore_others:true` means that all other topics from the `replay` module should be ignored. This allows precisely which part of the system should be replayed. The rules are deliberately generic (they could be applied to another module as well), but I didn't find another use-case yet.

## time handling
This point is still open. We want to be able to replay faster-than-real-time. For that we need to manipulate system time (`hrt_absolute_time()` and friends). I see two options:
- replay as fast as possible
- use a constant (maybe automatically set) speed-up factor

The first one is tricky, because we need to make sure no messages are lost. We could make orb synchronous, but this introduces lots of other problems such as deadlocks. So I tend towards the second. And then maybe report to the user, when the system starts to lose messages (i.e. replay is too fast).
A solution to this is basically independent from replay and highly overlaps with gazebo time simulation, so we should find a solution that works for both.
We need to identify which part of the system need the time, and in which way (like `hrt_absolute_time()`, clock, usleep, timeouts (like poll) etc). For some cases we don't want to speedup, for example status output every second or hardware access, like disk timeouts, and probably other cases.

## Workflow
A user wanting to replay has to:
- record the necessary topics
- setup orb publisher rules
- optionally setup parameter overrides
- optionally copy dataman mission file, if missions should be replayed
- `export replay=<abs_file_to_ulg>`
- build and run the target

The replay module has a `replay trystart` command, which checks if we want to replay (via env variable). If not in replay mode, it does nothing, so we can add it to the normal SITL startup script.
We could even use the same cmake target, but we should at least use a separate build directory for two reasons:
- the parameters from the replay file are stored in the SITL parameter's file. This can (and will) mess up normal SITL simulations.
- uorb rules should only be applied in replay mode.

The method demands one restriction to replayed topics: all additional timestamps except for the auto-generated field must be relative (because the replay module only knows about one timestamp). This is doable but requires a bit of refactoring (quite a few topics have additional timestamps). At the same time this allows to save some space, because relative timestamps don't need to be stored with 64bits.

Then, multi-topics are handled properly, but it is currently required that all or no instance is replayed. Should this restriction become a problem, we can still change it in the future.

We should also think about whether we can simply apply all parameters from the log file. Certain values for some parameters might not make sense for a simulated system.

## test results
I tested to replay ekf2. However the topics contain additional timestamps. I just set them to the same time as the published topic. This is wrong of course and has an effect to the result, but the following shows that it basically works:
Pixracer log:
![screenshot_ekf](https://cloud.githubusercontent.com/assets/281593/16042831/0069a080-323d-11e6-80ac-8f92bde72bae.png)
Replayed log:
![screenshot_ekf_replayed](https://cloud.githubusercontent.com/assets/281593/16042830/00675ffa-323d-11e6-8ee1-b4e9b41b7f78.png)
The replayed sensor values match exactly, and the estimated attitude at least look similar. We will have to see how it looks with proper timestamps.


If you want to test, you need to:
- add `#define ORB_USE_PUBLISHER_RULES` to `src/modules/uORB/uORBManager.hpp`
- setup publisher rules in `build_posix_sitl_default/src/firmware/posix/rootfs/orb_publisher.rules` file
- make sure the replayed topics contain no other timestamps, or fix them.
- `export replay=<abs_file_to_ulg>`
- run sitl
- `replay start`

Feedback, improvements & comments are welcome!
